src/main/java/com/spotify/docker/client/gcr/ContainerRegistryAuthSupplier.java

```diff
@@ -0,0 +1,251 @@
+
+    private final CredentialRefresher refresher = new DefaultCredentialRefresher();
+
+    public Builder(final GoogleCredentials credentials) {
+      this.credentials = credentials;
+    }
+
+    public Builder withScopes(Collection<String> scopes) {
+      this.scopes = scopes;
+      return this;
+    }
+
+    public Builder withMinimumExpiry(long duration, TimeUnit timeUnit) {
+      this.minimumExpiryMillis = TimeUnit.MILLISECONDS.convert(duration, timeUnit);
+      return this;
+    }
+
+    @VisibleForTesting
+    public Builder withClock(Clock clock) {
+      this.clock = clock;
+      return this;
+    }
+
+    public ContainerRegistryAuthSupplier build() throws IOException {
+      final GoogleCredentials credentials = this.credentials.createScoped(scopes);
+
+      // log some sort of identifier for the credentials, which requires looking at the
+      // instance type
+      if (credentials instanceof ServiceAccountCredentials) {
+        final String clientEmail = ((ServiceAccountCredentials) credentials).getClientEmail();
+        log.info("loaded credentials for service account with clientEmail={}", clientEmail);
+      } else if (credentials instanceof UserCredentials) {
+        final String clientId = ((UserCredentials) credentials).getClientId();
+        log.info("loaded credentials for user account with clientId={}", clientId);
+      }
+
+      return new ContainerRegistryAuthSupplier(credentials, clock, minimumExpiryMillis, refresher);
+    }
+  }
+
+  /**
+   * Refreshes a GoogleCredentials instance. This only exists for testing - we cannot mock calls to
+   * {@link GoogleCredentials#refresh()} as the method is final.
+   */
+  @VisibleForTesting
+  interface CredentialRefresher {
+
+    void refresh(GoogleCredentials credentials) throws IOException;
+  }
+
+  private static class DefaultCredentialRefresher implements CredentialRefresher {
+
+    @Override
+    public void refresh(final GoogleCredentials credentials) throws IOException {
+      credentials.refresh();
+    }
+  }
+
+  private final GoogleCredentials credentials;
+  // TODO (mbrown): change to java.time.Clock once on Java 8
+  private final Clock clock;
+  private final long minimumExpiryMillis;
+  private final CredentialRefresher credentialRefresher;
+
+  @VisibleForTesting
+  ContainerRegistryAuthSupplier(
+      final GoogleCredentials credentials,
+      final Clock clock,
+      final long minimumExpiryMillis,
+      final CredentialRefresher credentialRefresher) {
+
+    this.credentials = credentials;
+    this.clock = clock;
+    this.minimumExpiryMillis = minimumExpiryMillis;
+    this.credentialRefresher = credentialRefresher;
+  }
+
+  /**
+   * Get an accessToken to use, possibly refreshing the token if it expires within the
+   * minimumExpiryMillis.
+   */
+  private AccessToken getAccessToken() throws DockerException {
+    // synchronize attempts to refresh the accessToken
+    synchronized (credentials) {
```

Comment:
I don't think this will be an issue as I think when we call credentials.createScoped() we get a new instance, and only this class holds that instance. 

I don't like synchronizing on what seems like a shared object either but wasn't sure what else to use here to avoid redundant calls to refresh if there are concurrent calls to authFor - would be open to any suggestions you might have.